### PR TITLE
Improve UnionArray::is_nullable

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Relates to #6538.

# Rationale for this change
 

Always returning true is overly pessimistic, and whilst this is still not accurate, it is at least slightly more accurate

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-array/src/array/list_array.rs [hunk 1]
```diff
@@ -1181,4 +1181,18 @@ mod tests {
             .collect();
         assert_eq!(values, vec![Some(vec![1, 2, 3]), None, Some(vec![4, 5, 6])])
     }
+
+    #[test]
```
Comment (crepererum) [inline arrow-array/src/array/list_array.rs:1185]: I don't understand how this test connects to the code change. All the union members are nullable, no?

Comment (mbrobbel) [reply arrow-array/src/array/list_array.rs:1185]: This is a test from the linked issue https://github.com/apache/arrow-rs/issues/6538.

Comment (crepererum) [reply arrow-array/src/array/list_array.rs:1185]: I see, but it doesn't answer my question.

Comment (tustvold) [reply arrow-array/src/array/list_array.rs:1185]: The ListArray's field is not nullable, so doesn't permit a nullable child. Without the change in this PR the test fails

Comment (crepererum) [reply arrow-array/src/array/list_array.rs:1185]: Could we also have a more direct check for the code change, i.e. one that directly asserts `UnionArray::is_nullable` (in the respective module)? I'm fine with keeping the actual regression test.

**Review by findepi — APPROVED** [review]

File: arrow-array/src/array/list_array.rs (hunk 1 above)
Comment (alamb) [reply arrow-array/src/array/list_array.rs:1185]: I verified that this test fails without the code change:

```

called `Result::unwrap()` on an `Err` value: InvalidArgumentError("Non-nullable field of ListArray \"element\" cannot contain nulls")
thread 'array::list_array::tests::test_nullable_union' panicked at arrow-array/src/array/list_array.rs:228:54:
called `Result::unwrap()` on an `Err` value: InvalidArgumentError("Non-nullable field of ListArray \"element\" cannot contain nulls")
stack backtrace:
```

Comment (alamb) [reply arrow-array/src/array/list_array.rs:1185]: I also added a specific test in e78c87bff9

**Review by alamb — APPROVED** [review]
Thank you @tustvold

**Review by crepererum — COMMENTED** [review]
:shipit:

**Review by crepererum — APPROVED** [review]

**Comment by alamb** [comment]
Thank you everyone

**Pull request merged**